crates/bevy_render/src/render_event.rs

```diff
@@ -0,0 +1,121 @@
+use core::marker::PhantomData;
+
+use async_channel::{Receiver, Sender};
+use bevy_app::{App, Plugin, PreUpdate};
+use bevy_ecs::{
+    change_detection::MaybeLocation,
+    event::{Event, Events},
+    resource::Resource,
+    system::{Res, ResMut, SystemParam},
+};
+
+use crate::RenderApp;
+
+pub trait RenderEventApp {
+    fn add_render_event<E: Event>(&mut self) -> &mut Self;
+}
+
+impl RenderEventApp for App {
+    fn add_render_event<E: Event>(&mut self) -> &mut Self {
+        self.add_plugins(RenderEventPlugin::<E>::default())
+    }
+}
+
+struct RenderEventPlugin<E: Event>(PhantomData<E>);
+
+impl<E: Event> Default for RenderEventPlugin<E> {
+    fn default() -> Self {
+        Self(PhantomData)
+    }
+}
+
+impl<E: Event> Plugin for RenderEventPlugin<E> {
+    fn build(&self, app: &mut App) {
+        app.add_event::<E>()
+            .add_systems(PreUpdate, relay_render_events::<E>);
+    }
+
+    fn finish(&self, app: &mut App) {
+        let (sender, receiver) = async_channel::unbounded::<(E, MaybeLocation)>();
+
+        app.insert_resource(RenderEventReceiver(receiver));
+
+        let Some(render_app) = app.get_sub_app_mut(RenderApp) else {
+            return;
+        };
+
+        render_app.insert_resource(RenderEventSender(sender));
+    }
+}
+
+#[derive(Resource)]
+struct RenderEventReceiver<E: Event>(Receiver<(E, MaybeLocation)>);
+#[derive(Resource)]
+struct RenderEventSender<E: Event>(Sender<(E, MaybeLocation)>);
+
+fn relay_render_events<E: Event>(
+    mut events: ResMut<Events<E>>,
+    receiver: Res<RenderEventReceiver<E>>,
+) {
+    while let Ok((event, caller)) = receiver.0.try_recv() {
+        events.send_with_caller(event, caller);
+    }
+}
+
+/// An event writer for sending events from the render world to the main world.
+///
+/// Internally, this struct writes to a channel, the contents of which are relayed
+/// to the main world's [`Events`] stream during [`PreUpdate`]. Note that because
+/// the render world is pipelined, the events may not arrive before the next frame begins.
```

Comment:
They almost certainly will not, but they also could. I think we might want to be clearer about how this works. In practice, you'll receive events from rendering frame N-2 with respect the the main world, but that isn't guaranteed. Users should supply their own temporal correlation id if necessary.